src/routes/webhooks/apple.ts

```diff
@@ -57,6 +72,62 @@ interface AppleNotificationRequest {
   signedPayload: string;
 }
 
+const productIdToCycle = {
+  annualSpecial: SubscriptionCycles.Yearly,
+  annual: SubscriptionCycles.Yearly,
+  monthly: SubscriptionCycles.Monthly,
+};
+
+const renewalInfoToSubscriptionFlags = (
+  data: JWSRenewalInfoDecodedPayload,
+): UserSubscriptionFlags => {
+  const cycle =
+    productIdToCycle[data.autoRenewProductId as keyof typeof productIdToCycle];
+
+  if (isNullOrUndefined(cycle)) {
+    logger.error({ data }, '[Apple StoreKit]: Invalid auto renew product ID');
+    throw new Error('Invalid auto renew product ID');
+  }
+
+  return {
+    cycle,
+    subscriptionId: data.originalTransactionId,
+    createdAt: new Date(data.recentSubscriptionStartDate!),
+    expiresAt: new Date(data.renewalDate!),
+    provider: SubscriptionProvider.AppleStoreKit,
+  };
+};
+
+const getSubscriptionStatus = (
+  notificationType: ResponseBodyV2DecodedPayload['notificationType'],
+  subtype?: ResponseBodyV2DecodedPayload['subtype'],
+): UserSubscriptionStatus | null => {
+  switch (notificationType) {
+    case NotificationTypeV2.SUBSCRIBED:
+    case NotificationTypeV2.DID_RENEW:
+    case NotificationTypeV2.DID_CHANGE_RENEWAL_PREF: // Upgrade/Downgrade
+    case NotificationTypeV2.REFUND_REVERSED:
+      return UserSubscriptionStatus.Active;
+    case NotificationTypeV2.DID_CHANGE_RENEWAL_STATUS: // Disable/Enable Auto-Renew
+      return subtype === Subtype.AUTO_RENEW_ENABLED
+        ? UserSubscriptionStatus.Active
+        : UserSubscriptionStatus.Cancelled;
+    case NotificationTypeV2.DID_FAIL_TO_RENEW:
+      // When user fails to renew and there is no grace period
+      if (isNullOrUndefined(subtype)) {
+        return UserSubscriptionStatus.Cancelled;
+      }
+    case NotificationTypeV2.GRACE_PERIOD_EXPIRED:
+      return UserSubscriptionStatus.Cancelled;
+    case NotificationTypeV2.REFUND:
+    case NotificationTypeV2.EXPIRED:
+    case NotificationTypeV2.REVOKE: // We don't support Family Sharing, but to be on the safe side
+      return UserSubscriptionStatus.Expired;
+    default:
+      return null;
```

Comment:
does null mean we don't know how to handle this notificationType or sometimes it should be `null`?

Reply:
It means it's a notification type we don't know how to handle, and we'll log and throw an error

Reply:
ah I see you throw after evaluation, maybe better to throw from here directly? Then this method can also just have `UserSubscriptionStatus` as return type which is cleaner?

Reply:
Yeah, can do that.